File: evals/build_openings.py
"""Build server/openings.tsv (the ECO opening lookup table) from lichess-org/chess-openings.

Source: https://github.com/lichess-org/chess-openings (CC0). Each source row is
eco<TAB>name<TAB>pgn; we replay the PGN and key the opening by the position's EPD
(placement + turn + castling + en passant, via python-chess board.epd()) so the server
can look an opening up by position. Re-run to refresh:

    uv run --with chess python evals/build_openings.py
"""

import io
import pathlib
import urllib.request

import chess
import chess.pgn

BASE = "https://raw.githubusercontent.com/lichess-org/chess-openings/master"
FILES = ["a.tsv", "b.tsv", "c.tsv", "d.tsv", "e.tsv"]
OUT = pathlib.Path(__file__).parent.parent / "server" / "openings.tsv"


def main() -> None:
    rows = []
    for fname in FILES:
        text = (
            urllib.request.urlopen(f"{BASE}/{fname}", timeout=60).read().decode("utf-8")
        )
        for line in text.splitlines()[1:]:  # skip header
            if not line.strip():
                continue
            eco, name, pgn = line.split("\t")[:3]
            game = chess.pgn.read_game(io.StringIO(pgn))
            if game is None:
                continue
            board = game.end().board()
            rows.append((board.epd(), eco, name))

    # dedupe on EPD (first occurrence wins — sources are already ordered by depth)
    seen: set[str] = set()
    out_lines = ["epd\teco\tname"]
    for epd, eco, name in rows:
        if epd in seen:
            continue
        seen.add(epd)
        out_lines.append(f"{epd}\t{eco}\t{name}")

    OUT.write_text("\n".join(out_lines) + "\n", encoding="utf-8")
    print(f"wrote {OUT} ({len(out_lines) - 1} openings)")


if __name__ == "__main__":
    main()

File: apps/mcp-server/scripts/build-openings.mjs
/**
 * Generate data/openings.tsv (the ECO lookup table) from lichess-org/chess-openings (CC0).
 * Each source row is eco<TAB>name<TAB>pgn; we replay the PGN with chessops and key the opening
 * by positionKey (placement+turn+castling+ep) — the SAME key the lookup uses, so the table is
 * self-consistent and free of the python-chess-vs-chessops EPD parity risk. Re-run to refresh:
 *   node apps/mcp-server/scripts/build-openings.mjs
 */
import { Chess } from "chessops/chess";
import { parseSan } from "chessops/san";
import { makeFen } from "chessops/fen";
import { parsePgn } from "chessops/pgn";
import { writeFileSync, mkdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const BASE = "https://raw.githubusercontent.com/lichess-org/chess-openings/master";
const FILES = ["a.tsv", "b.tsv", "c.tsv", "d.tsv", "e.tsv"];
const positionKey = (fen) => fen.split(" ").slice(0, 4).join(" ");

const out = [];
let skipped = 0;
for (const f of FILES) {
  const text = await (await fetch(`${BASE}/${f}`)).text();
  for (const line of text.split("\n").slice(1)) {
    if (!line.trim()) continue;
    const [eco, name, pgn] = line.split("\t");
    if (!pgn) continue;
    const game = parsePgn(pgn)[0];
    if (!game) {
      skipped++;
      continue;
    }
    const pos = Chess.default();
    let node = game.moves;
    let okLine = true;
    while (node.children.length) {
      const child = node.children[0];
      const move = parseSan(pos, child.data.san);
      if (!move) {
        okLine = false;
        break;
      }
      pos.play(move);
      node = child;
    }
    if (!okLine) {
      skipped++;
      continue;
    }
    out.push(`${positionKey(makeFen(pos.toSetup()))}\t${eco}\t${name}`);
  }
}

const dataDir = join(dirname(fileURLToPath(import.meta.url)), "..", "data");
mkdirSync(dataDir, { recursive: true });
writeFileSync(join(dataDir, "openings.tsv"), out.join("\n") + "\n");
console.error(`wrote ${out.length} openings (${skipped} skipped) → data/openings.tsv`);
